File: lclang/src/main/java/lclang/libs/lang/classes/StringClass.java
package lclang.libs.lang.classes;

import lclang.Global;
import lclang.libs.lang.classes.numbers.IntClass;
import lclang.types.Type;
import lclang.types.Types;

import java.util.ArrayList;
import java.util.List;

public class StringClass extends LibraryClass {
    public static final String name = "string";
    public static final StringClass instance = new StringClass();
    public static final Type type = instance.classType;
    public String string;

    private StringClass() {
        super(name);
        constructor = method((caller, args) -> get(args.get(0).toString(caller)), Types.ANY, type);
    }

    private StringClass(String string) {
        this();
        this.string = string;
        globals.put("toString", method((caller, lcClasses) -> this, type));
        globals.put("length", IntClass.get(string.length()));
        globals.put("charAt", method((caller, args) -> CharClass.get(string.charAt(((IntClass) args.get(0)).value)),
            IntClass.TYPE, CharClass.type));
        globals.put("toArray", method((caller, lcClasses) -> {
            List<LCClass> classes = new ArrayList<>();

            char[] chars = string.toCharArray();
            for (char character: chars)
                classes.add(CharClass.get(character));

            return new ArrayClass(classes);
        }, ArrayClass.type));

        globals.put("lower", method((caller, args) -> StringClass.get(string.toLowerCase()), type));
        globals.put("upper", method((caller, args) -> StringClass.get(string.toUpperCase()), type));

        globals.put("split", method((caller, args) -> {
            List<LCClass> classes = new ArrayList<>();
            String[] parts = string.split(args.get(0).cast(StringClass.class).string);
            for (String part: parts)
                classes.add(StringClass.get(part));

            return new ArrayClass(classes);
        }, type, type));

        globals.put("substring", method((caller, args) ->
                        StringClass.get(
                            string.substring(
                                ((IntClass) args.get(0)).value,
                                ((IntClass) args.get(1)).value
                            )
                        ),
                IntClass.TYPE, IntClass.TYPE, type));

        globals.put("find", method((caller, args) ->
                        IntClass.get(string.indexOf(args.get(0).cast(CharClass.class).value)),
                CharClass.type, type));

        globals.put("findLast", method((caller, args) ->
                        IntClass.get(string.lastIndexOf(args.get(0).cast(CharClass.class).value)),
                CharClass.type, type));
    }

    public static StringClass get(String string) {
        if(Global.cashedClasses.containsKey(string))
            return (StringClass) Global.cashedClasses.get(string);

        StringClass clazz = new StringClass(string);
        Global.saveCache(string, clazz);
        return clazz;
    }
}

File: src/main/kotlin/lclang/lang/StringClass.kt
package lclang.lang

import lclang.LCClass
import lclang.Value
import lclang.constructor
import lclang.method
import lclang.methods.Method
import lclang.types.Types

class StringClass(): LCClass("string") {
    override var constructor: Method = constructor(listOf(Types.ANY), returnType = Types.STRING) { list ->
        StringClass(list[0].toString())
    }

    lateinit var string: String

    constructor(string: String): this() {
        this.string = string

        globals["length"] = Value(Types.INT, string.length)
        globals["charAt"] = method(listOf(Types.INT), Types.STRING) { string[it[0] as Int].toString() }
        globals["toArray"] = method(listOf(), Types.ARRAY) {
            ArrayClass().apply {
                for (c in string) add(Value(Types.CHAR, c))
            }
        }

        globals["toArray"] = method(listOf(Types.CHAR), Types.ARRAY) { args ->
            val parts = string.split(args[0] as Char)
            ArrayClass().apply {
                for(part in parts){
                    add(Value(Types.STRING, part))
                }
            }
        }

        globals["substring"] = method(listOf(Types.INT, Types.INT), Types.ARRAY) {
            string.substring(it[0] as Int, it[1] as Int)
        }

        globals["substring"] = method(listOf(Types.INT, Types.INT), Types.ARRAY) {
            string.substring(it[0] as Int, it[1] as Int)
        }

        globals["find"] = method(listOf(Types.CHAR), Types.INT) { args ->
            val needChar = (args[0] as CharClass).char
            string.indexOf(needChar)
        }
    }

    override fun equals(other: Any?): Boolean {
        if(other is StringClass)
            return other.string == string

        return false
    }

    override fun toString(): String = string
        .replace("\\n", "\n")
        .replace("\\r", "\r")
        .replace("\\b", "\b")
        .replace("\\t", "\t")
    override fun hashCode(): Int = string.hashCode()
}